sklearn/feature_extraction/tests/test_text.py

```diff
@@ -1196,3 +1197,46 @@ def build_preprocessor(self):
                                             .findall(doc),
                     stop_words=['and'])
     assert _check_stop_words_consistency(vec) is True
+
+
+@pytest.mark.parametrize('Estimator',
```

Comment:
I propose to refactor the test and break it in smaller tests which can be better parametrize. It solved the issue pointed out by @rth as well:

```python
@pytest.mark.parametrize('Estimator',
                         [CountVectorizer, TfidfVectorizer, HashingVectorizer])
@pytest.mark.parametrize(
    'input_type, err_type, err_msg',
    [('filename', FileNotFoundError, ''),
     ('file', AttributeError, "'str' object has no attribute 'read'")]
)
def test_callable_analyzer_error(Estimator, input_type, err_type, err_msg):
    data = ['this is text, not file or filename']
    with pytest.raises(err_type, match=err_msg):
        Estimator(analyzer=lambda x: x.split(),
                  input=input_type).fit_transform(data)


@pytest.mark.parametrize('Estimator',
                         [CountVectorizer, TfidfVectorizer, HashingVectorizer])
@pytest.mark.parametrize(
    'analyzer', [lambda doc: open(doc, 'r'), lambda doc: doc.read()]
)
@pytest.mark.parametrize('input_type', ['file', 'filename'])
def test_callable_analyzer_change_behavior(Estimator, analyzer, input_type):
    data = ['this is text, not file or filename']
    warn_msg = 'Since v0.21, vectorizer'
    with pytest.raises((FileNotFoundError, AttributeError)):
        with pytest.warns(ChangedBehaviorWarning, match=warn_msg) as records:
            Estimator(analyzer=analyzer, input=input_type).fit_transform(data)
    assert len(records) == 1
    assert warn_msg in str(records[0])


@pytest.mark.parametrize('Estimator',
                         [CountVectorizer, TfidfVectorizer, HashingVectorizer])
def test_callable_analyzer_reraise_error(tmpdir, Estimator):
    # check if a custom exception from the analyzer is shown to the user
    def analyzer3(doc):
        raise Exception("testing")

    f = tmpdir.join("file.txt")
    f.write("sample content\n")

    with pytest.raises(Exception, match="testing"):
        Estimator(analyzer=analyzer3, input='file').fit_transform([f])
```

@adrinjalali WDYT?

Reply:
Awesome! Thanks @glemaitre